src/data/media-service.ts

```diff
@@ -0,0 +1,99 @@
+import { decodeMediaData } from '~/utils/decode-media-data'
+import { VersionedApiService } from '~/data/api-service'
+
+import type {
+  ApiQueryParams,
+  DetailFromMediaType,
+  FrontendMediaType,
+  MediaResult,
+  MediaStoreResult,
+} from '~/store/types'
+import type { SupportedMediaType } from '~/constants/media'
+
+import type { AxiosResponse } from 'axios'
+
+class MediaService<T extends FrontendMediaType> {
+  private readonly mediaType: T
+
+  constructor(mediaType: T) {
+    this.mediaType = mediaType
+  }
+
+  /**
+   * Decodes the text data to avoid encoding problems.
+   * Also, converts the results from an array of media objects into an object with
+   * media id as keys.
+   * @param data - search result data
+   */
+  transformResults(data: MediaResult<T[]>): MediaStoreResult<T> {
+    const mediaResults = <DetailFromMediaType<T>[]>data.results
+    return {
+      ...data,
+      results: mediaResults.reduce((acc, item) => {
+        acc[item.id] = decodeMediaData<T>(item, this.mediaType)
+        return acc
+      }, {} as T extends 'audio' | 'image' ? Record<string, DetailFromMediaType<T>> : never),
+    }
+  }
+
+  /**
+   * Search for media items by keyword.
+   * @param params - API search query parameters
+   */
+  async search(
+    params: ApiQueryParams
+  ): Promise<MediaResult<Record<string, T>>> {
+    const res = await VersionedApiService.query<MediaResult<T[]>>(
+      <SupportedMediaType>this.mediaType,
+      params
+    )
+    console.log('res in search: ', res)
+    return this.transformResults(res.data)
+  }
+
+  /**
+   * Retrieve media details by its id.
+   * SSR-called
+   * @param params - the media id to fetch
+   */
+  async getMediaDetail(params: {
+    id: string
+  }): Promise<DetailFromMediaType<T>> {
+    if (!params.id) {
+      throw new Error(
+        `MediaService.getMediaDetail() id parameter required to retrieve ${this.mediaType} details.`
+      )
+    }
+
+    const res = await VersionedApiService.get<DetailFromMediaType<T>>(
+      <SupportedMediaType>this.mediaType,
```

Comment:
This style of casting isn't really used that much any more (I think because it's easy to confuse with JSX).

Why is the cast necessary though? `mediaType` already extends `'image' | 'audio'` which should match `SupportedMediaType` :thinking: 

```suggestion
      this.mediaType as SupportedMediaType,
```